// @progress/kendo-angular-dropdowns/dist/npm/common/taglist.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var tag_template_directive_1 = require("../common/templates/tag-template.directive");
var group_tag_template_directive_1 = require("../common/templates/group-tag-template.directive");
var util_1 = require("../common/util");
/**
 * @hidden
 */
var TagListComponent = /** @class */ (function () {
    function TagListComponent() {
        /**
         * A collection with the disabled tags' indices.
         */
        this.disabledIndices = new Set();
        this.removeTag = new core_1.EventEmitter();
    }
    TagListComponent.prototype.tagProp = function (tag, prop, index) {
        var propField = prop && this.getPropField(tag, prop, index);
        return util_1.getter(tag, propField);
    };
    TagListComponent.prototype.isTagDisabled = function (tag, positionIndex) {
        var _this = this;
        if (this.isGroupTag(tag)) {
            /** The `positionIndex` is used to determine after how many single tags is the group tag displayed =>
             * => it is used to increment the indices inside the group tag so that we determine the actual position index
             * of each group tag item as they appear in the `value` (important to check against `disabledIndices`)
             * e.g. `disabledIndices = [0, 1]` && `tags = ['Small', ['Medium', 'Large']]` => without the incrementation with
             * `positionIndex`, the group tag would yield [0, 1] as indices, while it should yield [1, 2]
             */
            return tag.every(function (_tag, index) { return _this.disabledIndices.has(index + positionIndex); });
        }
        return this.disabledIndices.has(positionIndex);
    };
    TagListComponent.prototype.deleteTag = function (event, tag, index) {
        event.preventDefault();
        event.stopImmediatePropagation();
        if (!this.disabled && event.which === 1) {
            this.removeTag.emit({ tag: tag, index: index });
        }
    };
    TagListComponent.prototype.itemId = function (tag, index) {
        if (tag) { //because of custom values
            return this.tagPrefix + "-" + this.tagProp(tag, this.valueField, index);
        }
    };
    TagListComponent.prototype.isGroupTag = function (tag) {
        return tag instanceof Array;
    };
    TagListComponent.prototype.tagAriaHidden = function (index) {
        return util_1.isPresent(this.focused) && this.focused !== index;
    };
    TagListComponent.prototype.getPropField = function (tag, prop, index) {
        // Needed for MultiSelectTree value binding (Heterogeneous Data)
        var fieldsCount = prop.length - 1;
        if (typeof prop === 'string') {
            return prop;
        }
        else if (this.valueDepth) {
            var depth = this.valueDepth[index];
            return fieldsCount < depth ? prop[fieldsCount] : prop[depth];
        }
        else {
            return prop.find(function (item) { return item in tag; });
        }
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], TagListComponent.prototype, "tags", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object)
    ], TagListComponent.prototype, "textField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], TagListComponent.prototype, "valueField", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array)
    ], TagListComponent.prototype, "valueDepth", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Number)
    ], TagListComponent.prototype, "focused", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", tag_template_directive_1.TagTemplateDirective)
    ], TagListComponent.prototype, "template", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", group_tag_template_directive_1.GroupTagTemplateDirective)
    ], TagListComponent.prototype, "groupTemplate", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], TagListComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], TagListComponent.prototype, "tagPrefix", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String)
    ], TagListComponent.prototype, "id", void 0);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Set)
    ], TagListComponent.prototype, "disabledIndices", void 0);
    tslib_1.__decorate([
        core_1.Output(),
        tslib_1.__metadata("design:type", core_1.EventEmitter)
    ], TagListComponent.prototype, "removeTag", void 0);
    TagListComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'kendo-taglist',
            template: "\n        <ul [attr.id]=\"id\" class=\"k-reset\">\n            <li\n                *ngFor=\"let tag of tags; let index = index;\"\n                role=\"option\"\n                [attr.id]=\"itemId(tag, index)\"\n                [attr.aria-hidden]=\"tagAriaHidden(index)\"\n                aria-selected=\"true\"\n                [attr.aria-setsize]=\"tags?.length\"\n                class=\"k-button\"\n                [ngClass]=\"{\n                    'k-state-focused': index === focused,\n                    'k-state-disabled': isTagDisabled(tag, index)\n                }\"\n            >\n                <ng-template *ngIf=\"isGroupTag(tag); then groupTag else singleTag\"></ng-template>\n                    <ng-template #groupTag>\n                        <span>\n                            <ng-template *ngIf=\"groupTemplate\"\n                                [templateContext]=\"{\n                                templateRef: groupTemplate.templateRef,\n                                $implicit: tag\n                            }\">\n                            </ng-template>\n                            <ng-template [ngIf]=\"!groupTemplate\">{{ tag.length }} {{ tag.length === 1 ? 'item' : 'items' }} selected</ng-template>\n                        </span>\n                    </ng-template>\n                    <ng-template #singleTag>\n                        <span>\n                            <ng-template *ngIf=\"template\"\n                                [templateContext]=\"{\n                                templateRef: template.templateRef,\n                                $implicit: tag\n                            }\">\n                            </ng-template>\n                            <ng-template [ngIf]=\"!template\">{{ tagProp(tag, textField, index) }}</ng-template>\n                        </span>\n                    </ng-template>\n\n                <span aria-label=\"delete\" [attr.aria-hidden]=\"index !== focused\" class=\"k-select\">\n                    <span class=\"k-icon k-i-close\" (mousedown)=\"deleteTag($event, tag, index)\">\n                    </span>\n                </span>\n            </li>\n        </ul>\n  "
        })
    ], TagListComponent);
    return TagListComponent;
}());
exports.TagListComponent = TagListComponent;
